plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,314 @@
+import java.lang.IllegalStateException
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ *
+ * The Viewport plugin allows to track objects on a map.
+ *
+ * It provides a structured approach to organizing camera management logic into states and transitions between them.
+ *
+ * at any given time, the viewport is either:
+ *  - idle (not updating the camera)
+ *  - in a state (camera is being managed by a ViewportState)
+ *  - transitioning (camera is being managed by a ViewportTransition)
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStatusObservers = CopyOnWriteArrayList<ViewportStatusObserver>()
+  private val internalStates = CopyOnWriteArraySet<ViewportState>()
+  private var currentCancelable: Cancelable? = null
+  private val transitions = HashMap<Pair<ViewportState?, ViewportState?>, ViewportTransition>()
+  private lateinit var delegateProvider: MapDelegateProvider
+  private lateinit var gesturesPlugin: GesturesPlugin
+  private val onMoveListener = object : OnMoveListener {
+    override fun onMoveBegin(detector: MoveGestureDetector) {
+      if (transitionsToIdleUponUserInteraction) {
+        currentCancelable?.cancel()
+        currentCancelable = null
+        updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_USER_INTERACTION)
+      }
+    }
+
+    override fun onMove(detector: MoveGestureDetector): Boolean {
+      // no-ops
+      return false
+    }
+
+    override fun onMoveEnd(detector: MoveGestureDetector) {
+      // no-ops
+    }
+  }
+
+  /**
+   * Returns list of registered states.
+   */
+  override val states: Set<ViewportState>
+    get() = internalStates.toSet()
+
+  /**
+   * Returns current [ViewportStatus].
+   *
+   * If current status is IDLE, returns ViewportStatus.State(null).
+   * @see addStatusObserver
+   */
+  override var status: ViewportStatus = ViewportStatus.State(null)
+    private set
+
+  private fun updateStatus(targetStatus: ViewportStatus, reason: String) {
+    val previousStatus = status
+    status = targetStatus
+    notifyStatusChanged(previousStatus, targetStatus, reason)
+  }
+
+  /**
+   * Executes a transition to requested state.
+   *
+   * When started, goes to [ViewportTransition]
+   * and to the final [ViewportState] when ended.
+   *
+   * If transition is canceled, state goes to IDLE.
+   *
+   * @param targetState The target [ViewportState] to transition to.
+   * @param completionListener The listener to observe the completion state.
+   */
+  override fun transitionTo(targetState: ViewportState, completionListener: CompletionListener?) {
+    if (status.getTargetState() === targetState) {
+      completionListener?.onComplete(false)
+      return
+    }
+    val fromState = status.getTargetState()
+    currentCancelable?.cancel()
+    currentCancelable = null
+    if (!states.contains(targetState)) {
```

Comment:
Could you please explain the logic behind this `internalStates` ? I've only seen that it's used here to throw if state was not added, is it the only motivation?

Reply:
The internal states was firstly introduced to hold a reference to the added states so that the users don't have to, however with the recent refactor, the users should be responsible to keep a reference to the states they want to use themselves. And thus the following APIs states/addState/removeState seems unnecessary, will consider to remove these 3 APIs. cc @macdrevx

Reply:
These 3 APIs are removed at https://github.com/mapbox/mapbox-maps-android/pull/1034